sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -883,6 +883,28 @@ createKernelArgMask(const pi::ByteArray &Bytes) {
   return Result;
 }
 
+static void handleXilinxKernels(RTDeviceBinaryImage* Img) {
+  static bool has_been_invoked = false;
+  const char* target = Img->getTarget();
+  const char* env_var = "XCL_EMULATION_MODE";
+  if (strncmp("fpga64_", target, 7) == 0 || strncmp("fpga32", target, 7) == 0) {
+    target += 7;
+    if (strcmp(target, "hw") == 0) {
+      if (has_been_invoked)
+        assert(!std::getenv(env_var));
+      ::unsetenv(env_var);
+    }
+    else {
+      if (has_been_invoked) {
+        assert(std::getenv(env_var));
+        assert(strcmp(std::getenv(env_var), target) ==0);
+      }
+      ::setenv(env_var, target, true);
```

Comment:
I guess it is useless to reset the value if you checked it before?
Or perhaps all this `has_been_invoked` optimization does not worth the effort?

Reply:
it is not an optimization it is a correctness check. this patch doesn't support mixing sw_emu with hw kernels in the same executable.